### ISSUE TITLE
Suggested changes to the assembly validator output interface

### ISSUE THREAD

**Issue by michaelbarton** [bioboxes/rfc#187]
In relation to #184, I was testing the assembly benchmarking interface with
quast. I initially ran:

```
biobox
  run \
  assembler_benchmark \
  bioboxes/quast \
  --input-fasta=contigs.fa \
  --output=output
```

This however failed because the argument `--input-ref` is required. I don't
have a reference genome for my contigs, however a directory is still required
even if it is empty. I next ran the command:

```
biobox
  run \
  assembler_benchmark \
  bioboxes/quast \
  --input-fasta=contigs.fa \
  --input-ref=ref \
  --output=metrics.txt
```

This produced the following error:

```
Traceback (most recent call last):
  File "/Users/michaelbarton/Library/Python/2.7/bin/biobox", line 9, in <module>
    biobox.run()
  File "/Users/michaelbarton/Library/Python/2.7/lib/python/site-packages/biobox_cli/main.py", line 30, in run
    util.select_module("command", opts["<command>"]).run(args)
  File "/Users/michaelbarton/Library/Python/2.7/lib/python/site-packages/biobox_cli/command/run.py", line 24, in run
    ctnr = bbx.run(argv)
  File "/Users/michaelbarton/Library/Python/2.7/lib/python/site-packages/biobox_cli/biobox.py", line 29, in run
    self.after_run(output, host_dst_dir)
  File "/Users/michaelbarton/Library/Python/2.7/lib/python/site-packages/biobox_cli/biobox_type/assembler_benchmark.py", line 52, in after_run
    self.copy_result_files(host_dst_dir, output)
  File "/Users/michaelbarton/Library/Python/2.7/lib/python/site-packages/biobox_cli/biobox_type/assembler_benchmark.py", line 26, in copy_result_files
    map(lambda f: shutil.move(os.path.join(biobox_output_dir,f), dst), output_files)
  File "/Users/michaelbarton/Library/Python/2.7/lib/python/site-packages/biobox_cli/biobox_type/assembler_benchmark.py", line 26, in <lambda>
    map(lambda f: shutil.move(os.path.join(biobox_output_dir,f), dst), output_files)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/shutil.py", line 299, in move
    copytree(src, real_dst, symlinks=True)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/shutil.py", line 177, in copytree
    os.makedirs(dst)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/os.py", line 157, in makedirs
    mkdir(name, mode)
OSError: [Errno 17] File exists: 'metrics'
```

Trying the same command with different output names produces the same error
with the corresponding name. Each time this created a file with the output
name. The command works correctly if the the output directly is created first
before running. When I run the command again the file `output/biobox.yaml` is
created with the contents:

``` yaml
version: 0.9.0
results:
  - name: HTML
    type: html
    inline: false
    value: combined_quast_output/standalone_report.html
    description: A summary of multiple metrics.
```

A text file with the QUAST metics for the assemly are in the file
`output/combined_quast_output/report.tsv`.
## Suggested changes
- The directory with reference genomes is optional. If this directory is not
  an argument, then the non-reference assembly metrics should be returned. I
  believe this follows the assembler benchmark signature where the fastq
  directory is optional.
- In the command line interface I think the `--output` flag should create a
  yaml file containing metrics of the benchmark evaluation instead of a
  directory with the quast generated files. I suggest this because the CLI is
  a simplied subset of using bioboxes for the most common cases. For example
  this is the case in the short_read_assembler CLI. This would also be a
  consistent UX when there are assembler benchmarks other than quast.

What do you think?

**Comment by michaelbarton** [comment]
To add additional context to these suggestions I have created [a recipe for
genome assembly][recipe] which includes using the assembly benchmark biobox. I
have made this as a likely user scenario and think if the `--input-ref` command
were optional and the `--output` flag instead produced a file this would
simplify the recipe.

[recipe]: https://github.com/bioboxes/bioboxes.org/blob/feature/genome-assembly-recipe/source/docs/assemble-a-genome.mkd

**Comment by pbelmann** [comment]
> The directory with reference genomes is optional. If this directory is not
   an argument, then the non-reference assembly metrics should be returned. I
   believe this follows the assembler benchmark signature where the fastq
   directory is optional.

I agree. It should be optional.

> In the command line interface I think the --output flag should create a
   yaml file containing metrics of the benchmark evaluation instead of a
   directory with the quast generated files. I suggest this because the CLI is
   a simplied subset of using bioboxes for the most common cases. For example
   this is the case in the short_read_assembler CLI. This would also be a
   consistent UX when there are assembler benchmarks other than quast.

If you want to remove the quast output directory, then I agree. But the output biobox.yaml file should follow the bioboxes evaluation format specification. 
If you want to add the tab separated file then you do it like this:

~~~YAML
version: 0.1.1
results:
  - name: HTML Summary
    description: html summary showing n50, ng50, .....
    value: report_standalone.html
    type: html
    inline: false
  - name: TSV metrics
    type: tsv
    inline: false
    description: html summary showing n50, ng50, ..... 
    value:  report.tsv
~~~

I have already included 4 different benchmarking bbxes in cami that are following interfaces that are not official bioboxes interafaces (not yet). And they are all using this format.
https://data.cami-challenge.org/analyseGetView  (Work in progress)

Is this ok Michael?

**Comment by michaelbarton** [comment]
My suggestion might be to make the TSV file a mandatory output for all evaluators, and then the CLI could mv this file to the specified `--output` location. How about that?

**Comment by pbelmann** [comment]
@michaelbarton 
How about we add to the evaluation format an optional tag called "cli"?
This way we could specify for the quast evaluation the following file:

~~~YAML
version: 0.1.1
cli: "TSV metrics"
results:
  - name: HTML Summary
    description: html summary showing n50, ng50, .....
    value: report_standalone.html
    type: html
    inline: false
  - name: TSV metrics
    type: tsv
    inline: false
    description: html summary showing n50, ng50, ..... 
    value:  report.tsv
~~~

The cli would parse the output yaml and move the tsv to the specified output location.
What do you think?

**Comment by michaelbarton** [comment]
My opinion is that the CLI should be and an easier method for using bioboxes, however the biobox.yaml should still be able to make sense without the CLI.

My suggestion for a mandatory TSV file was based on my thinking that each kind of evaluator would produce a series of key value metrics, and therefore we would always expect this.

We could call to discuss this if it would be faster in person.

**Comment by michaelbarton** [comment]
The assembly metrics output will be implemented in bioboxes/rfc#204. I will close this issue in favor of that one.
